### ISSUE TITLE
Cross-referencing components

### ISSUE THREAD

**Issue by krakz999** [goldbergyoni/nodebestpractices#799]
So the 1.1 best practice says:

> The very least you should do is create basic borders between components, assign a folder in your project root for each business component and make it self-contained - other components are allowed to consume its functionality only through its public interface or API.

Let's say I have an `users` component and an `orders` component. From the `orders` I have to access the **user DAL** or maybe a **service**. Would that be anti-pattern? Is the services layer considered as the interface between components? What is the recommended way to work with other components?

**Comment by kevynb** [comment]
Hello @krakz999,

That's always a tough question. I'll try to help you somehow. 

I would say that you need make a public api for your `orders` component to talk with your `users` component. If the service you want to use is the entrypoint to the `users` component then that's no problem. 
Talking to the *users DAL* through a service adds a layer of abstraction. If you need to add some business logic to the `users` side, you'll only have to to look into the `users` component. If you give the `orders` direct access to the `DAL` some `users` business logic might leak into the `orders`.

Are you storing your http controllers inside the component's folder or do you have another directory structure to store all those "infrastructure" files? If that's the case it's easier to expose the service publicly since it's already exposed to be used by the http controller (note: that's not a suggestion to change all your directory structure 😉 ).

@goldbergyoni what are your thoughts on that?

**Comment by goldbergyoni** [comment]
- Tweet


@krakz999 First, welcome 🙂

My thoughts - What @kevynb said. Also, we should set goals, what do we aim to (usually) achieve? Move fast. This is the holy grail that unifies most teams goals. Independent and reasonably-small component greatly supports this - If your Users component is small enough, you should feel very confident when adding more features/refactoring knowing that your change span only 1000 loc and not 10,000 loc (just an example). But, if your Users DAL is exposed to the entire world, any change might affect any other component, 10,000 loc, which might trigger fear, too many testing, and get you to move much slower.

Let's go with an example: Assume that you add a new DB column to users.

Option 1 🐢 - If you expose the DAL to collaborators, this might break Order who approaches this table without the new field, you'll have to conclude many collaborators before feeling confident. Not only that, these collaborators, will be able to access any of the component functions so the integration surface is huge. That will surely slow you down

Option 2 🦮 - Should you expose the service layer, you get a level of protection as you might provide a default value for the new field (within the service) and handle requests without the new field. But the service layer might be very rich, including many methods, and maintaining this layer to be compatible with many clients is not as easy as option 3'

Option 3 🏎 - Should you expose only HTTP API, the endpoints are very clear, documented, usually with fewer access doors (in a typical system there are less routes than internal functions) and in-overall the thinner integration layer one can provide to isolate a component from the world

Bottom line - Approaching the service is acceptable, exposing a network API only is much better

**Comment by torongu** [comment]
Hi krakz999. 
Just use CQS, it's a simple but powerful principle.
